[Refactor] 펀딩 도메인 아키텍처 구조 개선, 펀딩 생성 리펙토링

kdozlo opened the pull request:
## 반영 브랜치

- feat-funding -> main

<br/>

## 🔎 작업 내용

- Funding entity, domain 분리
  - FundingEntity, Funding
- repository 의존성 역전
  -  FundingRepository, FundingJPARepository, FundingRepositoryImpl
- mapper를 통한 도메인, dto간 의존성 완화
  - mapstruct 도입했으나 더 불편해져서 직접 구현으로 변경
     <img src="https://github.com/user-attachments/assets/5bd35ce3-4a82-449e-9aa7-8809c8f50bfc" width="400"/>
- 펀딩 도메인 아키텍처 구조 개선
```markdown
- controller/
  - FundingController.java
  - AnniversaryCategoryController.java
- service/
  - AnniversaryCategoryService.java
  - FundingPostService.java
  - FundingSchedulerService.java
  - FundingService.java
- mapper/
  - FundingMapper.java
- domain/
  - Funding.java
  - FundingDateAndStatus.java
  - FundingPrice.java
  - status/
    - FundingStatus.java(enum)
  - port/
    - FundingRepository.java(interface)
- infrastructure/
  - entity/
    - AnniversaryCategory.java
    - FundingEntity.java
  - repository/
    - AnniversaryCategoryRepository.java(JPA interface)
    - FundingJPARepository.java(JPA interface)
    - FundingRepositoryImpl.java
- dto
  - request/
    - DeleteFundingRequest.java
    - PostFundingRequest.java
  - response/
    - GetAnniversaryCategoryResponse.java
    - GetFundingCalendarResponse.java
    - GetFundingDetailResponse.java
    - GetFundingResponse.java
```
- 펀딩 생성 로직 리펙토링
  - **펀딩 생성 요청값 추가**
    - 펀딩생성 요청 날짜 추가
  - FundingDateAndStatus, FundingPrice 도메인 생성
    - 검증 도메인 로직 추가
- FundingPostService의 postFunding 메서드 중형 테스트, 단위 테스트 작성
  - 펀딩을_생성할_수_있다()
- 중형 테스트용 properites 추가
- 중형 테스트용 DB 추가, h2- 소형 테스트용 Fake repository 작성
  - FakeAnniversaryCategoryRepository
  - FakeFundingRepository
  - FakeProductOptionRepository
  - FakeProductOptionReposFakeProductRepositoryitory
- AnniversaryCategory 생성자 추가

<br/>

## 🚩 참고 사항

- 펀딩 생성 후 알림 로직 삭제
  - 추후에 개발 예정
- 펀딩 조회, 삭제 리펙토링 예정
- 펀딩 상태 변경 스케줄링 리펙토링 예정
- AnniversaryCategory 아키텍처 구조 개선 예정
- AnniversaryCategory, Consumer, Product, ProductOption 생성자에 id 추가

<br/>

kdozlo committed:
[Refactor] 펀딩 생성 기능 리펙토링

- entity, domain 분리
- repository 의존성 역전
- mapper 패키지 생성을 통한 도메인의 dto 의존성 삭제
  - mapstruct 사용
- FundingDateAndStatus, FundingPrice 도메인 생성
  - 검증 도메인 로직 추가
- 펀딩 생성 후 알림 로직 삭제
  - 추후에 개발 예정

kdozlo committed:
[Rename] GetFriendCommand -> GetFriendResponse

kdozlo committed:
[Chore] mapstruct 삭제

- mapper 직접 구현으로 변경

kdozlo committed:
[Refactor] 제품, 제품 옵션 일치 확인 로직 수정

kdozlo committed:
[Test] FundingPostService의 postFunding 메서드 중형 테스트, 단위 테스트 작성

- 펀딩을_생성할_수_있다()
- 중형 테스트용 properites 추가
- 중형 테스트용 DB 추가, h2
- 소형 테스트용 Fake repository 작성
  - FakeAnniversaryCategoryRepository
  - FakeFundingRepository
  - FakeProductOptionRepository
  - FakeProductOptionReposFakeProductRepositoryitory
- AnniversaryCategory 생성자 추가
- 펀딩 생성 요청값 추가
  - 펀딩생성 요청 날짜 추가
- AnniversaryCategory, Consumer, Product, ProductOption 생성자에 id 추가

kdozlo committed:
[Style] 불필요 import문 삭제

limsubinn commented on BE/src/main/java/com/d201/fundingift/funding/domain/FundingDateAndStatus.java:
```diff
@@ -0,0 +1,75 @@
+    private FundingStatus fundingStatus;
+
+    @Builder
+    public FundingDateAndStatus(LocalDate anniversaryDate, LocalDate startDate, LocalDate endDate, FundingStatus fundingStatus) {
```
public을 의도한게 맞으실까요?!

kdozlo replied:
private인데 놓쳤네요. 감사합니다

limsubinn commented on BE/src/main/java/com/d201/fundingift/funding/domain/FundingDateAndStatus.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    //시작일이 오늘이면 IN_PROGRESS로 상태 변경, 미래면 PRE_PROGRESS
+    private static FundingStatus IsStartDateToday(LocalDate startDate) {
```
메서드명 첫글자 소문자로 수정하면 좋을 것 같습니다

kdozlo replied:
헉 수정했습니다.

limsubinn commented on BE/src/main/java/com/d201/fundingift/funding/intrastructure/entity/FundingEntity.java:
```diff
@@ -77,10 +73,9 @@ public class Funding extends BaseTime {
     @Column(nullable = false)
     @ColumnDefault("'PRE_PROGRESS'")
-    private FundingStatus fundingStatus;
+    private String fundingStatus;
```
enum 타입에서 String으로 변경한 이유가 있을까요?

kdozlo replied:
이 부분은 아직 고민중입니다.
기존 방식인 엔티티 + 도메인에서 엔티티, 도메인으로 분리하니 FundingStatus가 애매해졌어요. 세가지 방법이 있습니다.
1. FundingStatus도 엔티티, 도메인으로 각각 분리하자.
2. FundingStatus를 엔티티쪽에서 가지고 있는다.
3. FundingStatus를 도메인쪽에서 가지고 있는다.

1번 방식의 경우, 각각 enum을 만들어줘야 하는데 단순 enum이라 코드 중복이 생기는 상황입니다.

2번 방식의 경우, 도메인쪽에서는 상태를 String으로 받아야합니다. 하지만 도메인이 서비스의 정책을 결정하고, 가장 중심 역할인데 상태에 대한 제한을 걸지 않으면 안된다고 생각해서 안된다고 생각했습니다.
만약 FundingStatus를 엔티티쪽에서 가지고 있고, 이를 도메인에서 쓴다면 도메인이 엔티티를 의존하는 구조가 되어 현재 엔티티 도메인 분리라는 목표에 부합하지 않는다고 생각합니다.

3번 방식의 경우, 도메인쪽에서 FundingStatus를 쓰고, 데이터베이스의 엔티티에서는 String으로 받는 방식입니다. 도메인이 서비스의 핵심인 만큼 이 방법이 가장 적합하다고 판단했습니다.
만약 FundingStatus를 엔티티쪽에서 가저와 쓴다면, 캡슐화에 위배된다고 생각해서 일단은 String으로 받기로 했습니다.

현재로서는 3버 방식이 가장 적합하다고 생각하곤 있지만 더 좋은 방식이 있으면 바꿀 생각이 있습니다.

limsubinn replied:
답변 감사합니다! 👍

limsubinn commented on BE/src/main/java/com/d201/fundingift/funding/dto/request/PostFundingRequest.java:
```diff
@@ -15,6 +17,9 @@ public class PostFundingRequest {
     private Integer targetPrice;
 
+    @Schema(description = "펀딩생성 요청 날짜", example = "2024-12-12")
+    private LocalDate currentDate;
```
펀딩 생성 요청 날짜를 넣은 이유가 있을까요?
제 생각에는 프론트에서 요청을 위조할 수 있어 바람직하지 않아 보입니다.

kdozlo replied:
현재 생성 요청 날짜보다 펀딩 시작일이 과거인 경우 예외 로직이 있습니다.

요청값에서 펀딩 생성 요청 날짜를 안 줬을 때, 소비자가 1월 31일에 펀딩 시작일이 1월 31일인 펀딩을 생성을 요청했다고 가정해 볼게요.
그런데 서버에서 해당 요청을 2월 1일에 처리한다고 하면, 소비자가 1월 31일에 해당 요청을 보냈더라도 
서버 내부에서는  생성 요청 날짜와 펀딩 시작일 검증 로직에서 생성 요청 날짜가 2월 1일이 되고, 펀딩 시작일이 1월 31일이 되어 펀딩 생성 실패가 되는 경우가 생깁니다.

이 상황을 고려해서 요청값에 펀딩 생성 요청 날짜를 넣었는데, 그 문제보다 요청 위조가 더 위험하겠네요. 이부분도 다시 원래대로 수정 하고, 위의 문제에 대한 해결 방법은 좀 더 생각해 보겠습니다.

limsubinn replied:
말씀하신 부분이 날짜가 넘어가면서 생기는 예외 같은데, 이 경우에는 사용자가 응답을 받았을 때 이미 12시가 지난 후니까... 펀딩 생성 요청을 해당 날짜로 다시 하는게 맞지 않을까요....?!?!

kdozlo replied:
근데 생각해보니 그게 맞네요. 너무 과하게 생각해서 놓친 부분이 있네요.

limsubinn commented on BE/src/main/java/com/d201/fundingift/friend/service/FriendService.java:
```diff
@@ -103,8 +103,8 @@ public List<GetFriendStoryResponse> getFriendsStory() {
             //친구의 펀딩 목록 중 진행중이고 시작일이 제일 빠른 하나 반환
-            List<Funding> privateFundings = getAllByConsumerIdAndFundingStatusAndIsPrivateAndDeletedAtIsNullOrderByStartDateAsc(f, true);
-            List<Funding> notPrivateFundings = getAllByConsumerIdAndFundingStatusAndIsPrivateAndDeletedAtIsNullOrderByStartDateAsc(f, false);
+            List<FundingEntity> privateFundings = getAllByConsumerIdAndFundingStatusAndIsPrivateAndDeletedAtIsNullOrderByStartDateAsc(f, true);
+            List<FundingEntity> notPrivateFundings = getAllByConsumerIdAndFundingStatusAndIsPrivateAndDeletedAtIsNullOrderByStartDateAsc(f, false);
```
이렇게 쿼리를 두 번 날리는 것과 한꺼번에 조회 후 애플리케이션 단에서 필터링 하는 로직 중 어떤 것이 더 나을까요? 혹시 비교해보셨나요?

![image](https://github.com/user-attachments/assets/35f93881-8393-44a4-8f4a-a09390a58ea2)

kdozlo replied:
이부분은 리펙토링 전 친구 데이터들이 redis에 있어서 저렇게 나누어져 있었고, 
현재는 RDB가 되었기 때문에 추후 리펙토링 과정에서 수정 예정입니다.

limsubinn commented on BE/src/main/java/com/d201/fundingift/funding/intrastructure/repository/FundingJPARepository.java:
```diff
@@ -0,0 +1,74 @@
+            "AND f.consumer.id = :consumerId AND f.isPrivate = false AND f.deletedAt IS NULL")
+    List<FundingEntity> findAllByConsumerIdAndIsPrivateAndDeletedAtIsNull(@Param("consumerId") Long consuerId, @Param("year") Integer year, @Param("month") Integer month);
+
+    @Query("select p FROM funding f right join f.product p " +
+            "where p.status = 'ACTIVE' and p.deletedAt is null " +
+            "group by p order by count(f) desc")
+    Slice<Product> findProductSliceOrderByFundingCount(Pageable pageable);
```
이 부분 제가 제거했으니 제거해주시면 될 것 같습니다!!

kdozlo replied:
넵넵 머지할 때 반영하겠습니다.

limsubinn reviewed (commented):
아키텍처 구조를 제가 잘 몰라서 코드 보기가 조금 어렵네요.. 우선 눈에 보이는 것들만 코멘트 달았어요..! 코드 참고해서 아키텍처 구조랑 테스트 코드 작성 공부해보겠습니다 ㅎㅎ 고생하셨습니다!!

kdozlo commented:
> 아키텍처 구조를 제가 잘 몰라서 코드 보기가 조금 어렵네요.. 우선 눈에 보이는 것들만 코멘트 달았어요..! 코드 참고해서 아키텍처 구조랑 테스트 코드 작성 공부해보겠습니다 ㅎㅎ 고생하셨습니다!!

자세한 코드 리뷰 감사합니다. 고민했던 부분 그대로 캐치해서 달아주셨네요. 감사합니다.

kdozlo committed:
[Fix] 펀딩 생성 요청값 수정

- 펀딩생성 요청 날짜 제거
  - PR 피드백 반영

kdozlo committed:
[Chore] yml 파일 수정

kdozlo committed:
[Fix] 충돌 해결

- @Entity(name = "funding") -> @Entity
                               @Table(name = "funding") 수정
- @Entity(name = "Friend") -> @Entity
                              @Table(name = "friend") 수정
- FundingiftApplicationTests 삭제
